Tests/NIOTests/ChannelTests.swift

```diff
@@ -1716,4 +1716,143 @@ public class ChannelTests: XCTestCase {
+        defer {
+            XCTAssertNoThrow(try group.syncShutdownGracefully())
+        }
+        let serverEL = group.next()
+        let clientEL = group.next()
+        precondition(serverEL !== clientEL)
+        let sc = try SocketChannel(eventLoop: clientEL as! SelectableEventLoop, protocolFamily: PF_INET)
+
+        class WriteImmediatelyHandler: ChannelInboundHandler {
+            typealias InboundIn = Any
+            typealias OutboundOut = ByteBuffer
+
+            private let writeDonePromise: EventLoopPromise<Channel>
+
+            init(writeDonePromise: EventLoopPromise<Channel>) {
+                self.writeDonePromise = writeDonePromise
+            }
+
+            func channelActive(ctx: ChannelHandlerContext) {
+                var buffer = ctx.channel.allocator.buffer(capacity: 4)
+                buffer.write(string: "foo")
+                ctx.writeAndFlush(NIOAny(buffer)).map {
+                    ctx.channel
+                }.cascade(promise: self.writeDonePromise)
+            }
+        }
+
+        let serverWriteHappenedPromise: EventLoopPromise<Channel> = serverEL.next().newPromise()
+        let clientHasRegistered: EventLoopPromise<()> = serverEL.next().newPromise()
+        let clientHasUnregistered: EventLoopPromise<()> = serverEL.next().newPromise()
+
+        let bootstrap = try ServerBootstrap(group: serverEL)
+            .childChannelInitializer { channel in
+                channel.pipeline.add(handler: WriteImmediatelyHandler(writeDonePromise: serverWriteHappenedPromise))
+            }
+            .bind(host: "127.0.0.1", port: 0).wait()
+
+        // This is a bit ugly, we're trying to fabricate a situation that can happen in the real world which is that
+        // a socket is readable straight after becoming registered & connected.
+        // In here what we're doing is that we flip the order around and connect it first, make sure the server
+        // has written something and then on registration something is available to be read. However it should not
+        // be read yet (and in this case never will be) as we haven't send channelActive yet (which will never happen)
+        // in this example.
+        _ = try sc.selectable.connect(to: bootstrap.localAddress!)
+        let acceptedChannel = try serverWriteHappenedPromise.futureResult.wait()
+        try sc.pipeline.add(handler: ReadDoesNotHappen(hasRegisteredPromise: clientHasRegistered,
+                                                       hasUnregisteredPromise: clientHasUnregistered)).then {
+                // this will succeed and should not cause the socket to be read even though there'll be something
+                // available to be read immediately
+                sc.register()
+            }.then {
+                // this should fail (and will on Darwin but _not_ Linux) as the socket is already pre-connected ;)
+                // on Linux a subsequent connect will just silently succeed, go figure...
```

Comment:
This really makes me sad :( I wonder if there is no other way to test this in a more "consistent" way

Reply:
@normanmaurer after @Lukasa 's patch is in I can create a `Socket` subclass which lets a subsequent 'special' connect succeed on all platforms. That way we can test even more as the channel does become active.

Reply:
@weissi which patch you are talking about ? :)

Reply:
@normanmaurer I must have dreamt ;)